alica_engine/include/engine/RunnableObjectNew.h

```diff
@@ -0,0 +1,131 @@
+
+#include <atomic>
+#include <functional>
+#include <memory>
+#include <string>
+
+namespace alica
+{
+namespace test
+{
+class TestContext;
+}
+class RunningPlan;
+class AlicaEngine;
+class IAlicaWorldModel;
+
+class TraceRunnableObject
+{
+public:
+    TraceRunnableObject()
+            : _tracingType(TracingType::DEFAULT)
+            , _runTraced(false)
+    {
+    }
+
+    enum class TracingType : uint8_t
+    {
+        // Use the trace context of the lowest ancestor plan for which tracing is not skipped
+        DEFAULT,
+        // Skip tracing
+        SKIP,
+        // Create a root trace
+        ROOT,
+        // Provide a custom trace context
+        CUSTOM
+    };
+
+    IAlicaTrace* getTrace() const { return _trace ? _trace.get() : nullptr; };
+
+    // Set the tracing type for this runnable object. customTraceContextGetter is required for custom tracing
+    // & this method will be called to get the parent trace context before initialiseParameters is called
+    void setTracing(TracingType type, std::function<std::optional<std::string>()> customTraceContextGetter = {});
+    void setupTraceContext(const std::string& name, RunningPlan* rp, const IAlicaTraceFactory* traceFactory);
+    void cleanupTraceContext();
+    void traceRunCall();
+
+private:
+    TracingType _tracingType;
+    std::function<std::optional<std::string>()> _customTraceContextGetter;
+    std::unique_ptr<IAlicaTrace> _trace;
+    // True if the behaviour/plan's run method has already been logged in the trace
+    bool _runTraced;
+    IAlicaTraceFactory* _traceFactory;
+};
+
+/**
+ * The base class for BasicBehaviour and BasicPlan
+ */
+class RunnableObjectNew
+{
+protected:
+    using TracingType = TraceRunnableObject::TracingType;
+
+    RunnableObjectNew(IAlicaWorldModel* wm, const std::string& name = "");
+    virtual ~RunnableObjectNew() = default;
+    // virtual ~RunnableObjectNew() { std::cout << _name << " destructor" << std::endl; }
+
+    static constexpr int DEFAULT_MS_INTERVAL = 100;
+
+    virtual void doInit() = 0;
+    virtual void doRun() = 0;
+    virtual void doTerminate() = 0;
+
+    void setTracing(TracingType type, std::function<std::optional<std::string>()> customTraceContextGetter = {})
+    {
+        _runnableObjectTracer.setTracing(type, customTraceContextGetter);
+    }
+    const std::string& getName() { return _name; };
+    IAlicaTrace* getTrace() const { return _runnableObjectTracer.getTrace(); };
+
+    void sendLogMessage(int level, const std::string& message) const;
+    RunningPlan* getPlanContext() const { return _runningplanContext; }
+    const std::shared_ptr<Blackboard> getBlackboard() { return _blackboard; }
+    IAlicaWorldModel* getWorldModel() { return _wm; };
+
+    void start(RunningPlan* rp);
+    void stop();
+    void setEngine(AlicaEngine* engine) { _engine = engine; };
+    // Only plan will have these
+    void addKeyMapping(int64_t wrapperId, const KeyMapping* keyMapping);
```

Comment:
Will these not end up in the derived class instead?

Reply:
They will probably move to the RunningPlan. Will refactor in the 3rd cleanup pr.